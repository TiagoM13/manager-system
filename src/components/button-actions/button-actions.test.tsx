import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import { ButtonActions, ButtonActionsProps } from '.';

describe('<ButtonActions/>', () => {
  const onEditMock = jest.fn();
  const onDeleteMock = jest.fn();

  const renderComponent = (props: Partial<ButtonActionsProps> = {}) => {
    return render(
      <ButtonActions onEdit={onEditMock} onDelete={onDeleteMock} {...props} />,
    );
  };

  test('should render correctly component', () => {
    const { getByTestId, queryByText, container, asFragment } =
      renderComponent();

    expect(getByTestId('icon-btn')).toBeInTheDocument();

    const button = getByTestId('btn-actions');
    fireEvent.click(button);

    expect(container.querySelector('.hidden')).not.toBeInTheDocument();
    expect(queryByText('Editar')).toBeInTheDocument();
    expect(queryByText('Deletar')).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });

  test('should display the options menu when the button is clicked', () => {
    const { getByText, getByTestId, container } = renderComponent();

    fireEvent.click(getByTestId('btn-actions'));

    expect(container.querySelector('.block')).toBeInTheDocument();

    expect(getByText('Editar')).toBeInTheDocument();
    expect(getByText('Deletar')).toBeInTheDocument();
  });

  test('should call onEdit when the edit button is clicked', () => {
    const { getByTestId, getByText, container } = renderComponent();

    fireEvent.click(getByTestId('btn-actions'));

    fireEvent.click(getByText('Editar'));

    expect(container.querySelector('.hidden')).toBeInTheDocument();
    expect(onEditMock).toHaveBeenCalledTimes(1);
  });

  test('should call onDelete when the delete button is clicked', () => {
    const { getByTestId, getByText, container } = renderComponent();

    fireEvent.click(getByTestId('btn-actions'));

    fireEvent.click(getByText('Deletar'));

    expect(container.querySelector('.hidden')).toBeInTheDocument();
    expect(onDeleteMock).toHaveBeenCalledTimes(1);
  });

  test('should close the menu when clicking outside of the component', () => {
    const { getByTestId, container } = renderComponent();

    fireEvent.click(getByTestId('btn-actions'));

    fireEvent.mouseDown(document);

    expect(container.querySelector('.hidden')).toBeInTheDocument();
  });
});
