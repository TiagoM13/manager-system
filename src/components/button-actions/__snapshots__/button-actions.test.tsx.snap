// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonActions/> should render correctly component 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <button
      class="bg-white hover:bg-slate-200/50 border border-slate-400/50 rounded-md p-1.5 duration-300 disabled:bg-slate-300/50 disabled:hover:bg-slate-300/50 outline-slate-400"
      data-testid="btn-actions"
    >
      <svg
        class="size-4 text-slate-800"
        data-testid="icon-btn"
        fill="currentColor"
        height="1em"
        viewBox="0 0 256 256"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M156,128a28,28,0,1,1-28-28A28,28,0,0,1,156,128ZM48,100a28,28,0,1,0,28,28A28,28,0,0,0,48,100Zm160,0a28,28,0,1,0,28,28A28,28,0,0,0,208,100Z"
        />
      </svg>
    </button>
    <div
      class="block bg-white shadow-sm shadow-slate-400 w-28 absolute top-8 -left-20 rounded-lg z-20"
    >
      <div
        class="py-1"
      >
        <button
          class="flex items-center gap-4 w-full p-2 hover:bg-sky-100"
        >
          <svg
            class="size-4 text-sky-600"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M230.14,70.54,185.46,25.85a20,20,0,0,0-28.29,0L33.86,149.17A19.85,19.85,0,0,0,28,163.31V208a20,20,0,0,0,20,20H92.69a19.86,19.86,0,0,0,14.14-5.86L230.14,98.82a20,20,0,0,0,0-28.28ZM91,204H52V165l84-84,39,39ZM192,103,153,64l18.34-18.34,39,39Z"
            />
          </svg>
          <span
            class="text-sm text-sky-600 text-wrap"
          >
            Editar
          </span>
        </button>
        <button
          class="flex items-center gap-4 w-full p-2 hover:bg-red-100"
        >
          <svg
            class="size-4 text-red-600"
            fill="currentColor"
            height="1em"
            viewBox="0 0 256 256"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M216,48H180V36A28,28,0,0,0,152,8H104A28,28,0,0,0,76,36V48H40a12,12,0,0,0,0,24h4V208a20,20,0,0,0,20,20H192a20,20,0,0,0,20-20V72h4a12,12,0,0,0,0-24ZM100,36a4,4,0,0,1,4-4h48a4,4,0,0,1,4,4V48H100Zm88,168H68V72H188ZM116,104v64a12,12,0,0,1-24,0V104a12,12,0,0,1,24,0Zm48,0v64a12,12,0,0,1-24,0V104a12,12,0,0,1,24,0Z"
            />
          </svg>
          <span
            class="text-sm text-red-600 text-wrap"
          >
            Deletar
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
